### ISSUE TITLE
jpm run: do temporary extension install

### ISSUE THREAD

**Issue by freaktechnik** [mozilla-jetpack/jpm#468]
It seems that currently JPM isn't installing extensions as "temporary" yet, when supported. From Firefox 45 and onwards unsigned extensions can be temporarily installed (for example with about:debugging). Even though the extension ran with jpm run appears in about:debugging, it still gets disabled when resetting the `xpinstall.signatures.required` pref back to true.

In order for JPM to continue working with the Firefox 46 Beta or release and later.

**Comment by kumar303** [comment]
How would this be helpful for development with jpm? It already installs an add-on in a temporary profile which achieves the same effect.

Anyway, I'd be more interested in adding new features like this to [web-ext](https://github.com/mozilla/web-ext) rather than jpm.

**Comment by freaktechnik** [comment]
So you could still use release builds, since from what I understand about:debugging will also be available on release builds.

I understand that for you this is more interesting to implement in web-ext, however jetpack is at least still supported until XPCOM is artificially deprecated from what I understand.

**Comment by kumar303** [comment]
Yes, `jpm` is still supported so I will leave this issue open for it. I would like to minimize the amount of duplicated effort between web-ext and jpm though, if possible!

As for the use case of testing extensions in release builds, I can see how this feature will be useful.

**Comment by freaktechnik** [comment]
I guess it would be best if fx-runner could do this, though you might not want to use that in the web-ext tool, but that's not up to me.

**Comment by kumar303** [comment]
Yes, I am using `fx-runner` in web-ext currently but so far I have had to replace its `addExtension()` function -- it doesn't support web extensions. My plan is to use as much of it as I can so if someone patches it for temporary extensions then that would be nice :)

**Comment by kumar303** [comment]
> I have had to replace its addExtension() function 

I might try to uplift this patch. It's still in the "bang on everything until it works" stage.

**Comment by freaktechnik** [comment]
This will become an issue from the 2nd of August on, since then Release and Beta will both no longer have the pref. This mean the travis tests will break until either this is fixed or the used Firefox build is changed.

**Comment by kumar303** [comment]
does temporary installation even work on Release and Beta? Doesn't the signature restriction still apply?

**Comment by kumar303** [comment]
It may be better to use [unbranded builds](https://wiki.mozilla.org/Addons/Extension_Signing#Unbranded_Builds) for jpm development.

**Comment by freaktechnik** [comment]
about:debugging works on release and beta.

**Comment by kumar303** [comment]
ok, cool, I guess temporary installation is allowed in Release.

So, yeah, one solution here would be to make jpm do what web-ext does: connect to the remote debugger and [use](https://github.com/mozilla/web-ext/blob/master/src/firefox/remote.js#L66) the `installTemporaryAddon()` actor function. However, this remote function is only supported in Firefox 49. You could extract that remote firefox code into its own module that web-ext and jpm could share. This would make maintenance easier. It could also be proposed upstream to the [firefox-profile](https://www.npmjs.com/package/firefox-profile) module but I don't know how active that is.

**Comment by carlin-q-scott** [comment]
So I can't test my add-on locally anymore because of this issue, at least not against the version of Firefox that my users will have. I also had to pin my travis-ci build to Firefox v47 to avoid it failing.

Is anyone working on implementing the suggested fix? As a temporary solution, couldn't the generated Firefox profile have xpinstall.signatures.required=false?

**Comment by freaktechnik** [comment]
@carlin-q-scott The unbranded releases are pretty much equal to the released Firefox, just that they still allow you to disable signing: https://blog.mozilla.org/addons/2016/07/29/extension-signing-availability-of-unbranded-builds/
